src/main/java/com/google/api/generator/gapic/protoparser/HttpRuleParser.java

```diff
@@ -99,40 +97,67 @@
       }
     }
 
-    return Optional.of(bindings);
+    // TODO: support nested message fields bindings
+    String body = httpRule.getBody();
+    Set<String> bodyParameters;
+    Set<String> queryParameters;
+    if (!inputMessageOpt.isPresent()) {
+      // Must be a mixin, do not support full HttpRuleBindings for now
+      bodyParameters = ImmutableSet.of();
+      queryParameters = ImmutableSet.of();
+    } else if (Strings.isNullOrEmpty(body)) {
+      bodyParameters = ImmutableSet.of();
+      queryParameters = Sets.difference(inputMessageOpt.get().fieldMap().keySet(), bindings);
+    } else if (body.equals(ASTERISK)) {
+      bodyParameters = Sets.difference(inputMessageOpt.get().fieldMap().keySet(), bindings);
+      queryParameters = ImmutableSet.of();
+    } else {
+      bodyParameters = ImmutableSet.of(body);
+      Set<String> bodyBinidngsUnion = Sets.union(bodyParameters, bindings);
+      queryParameters =
+          Sets.difference(inputMessageOpt.get().fieldMap().keySet(), bodyBinidngsUnion);
+    }
+
+    HttpRuleBindings.Builder httpBindings = HttpRuleBindings.builder();
+    httpBindings.setHttpVerb(httpRule.getPatternCase().toString());
+    httpBindings.setPattern(pattern);
```

Comment:
What about a version of `setPattern` that takes a `PatternCase`? Then we can move `getPattern` into `HttpRuleBindings`, where it fits better anyway.

Reply:
In that case, I think the whole parsing logic must be moved to `HttpRuleBindings`. I.e. the current architecture has a split between the class which simply contains the values (`HttpRuleBindings`, a simple POJO, javabean, whatever), and `HttpRuleParser` - one of potential classes, which knows how to populate `HttpRuleBinding` class from a HttpRule. So basically `HttpRuleBindings` class is detached from HttpRule per se, and may be populated from another source by some other sort of parser or manually. From my experience splitting this stuff into two classes (one contains values, the other one (or a set of, one for each source) knows how to populate those values from a source) usually works better in a long run (especially if a persistence layer is used, like a database, though this is irrelevant here). I can merge them, if you wish, though, I don't have a strong preference here.